containers/julia/.devcontainer/devcontainer.json

```diff
@@ -0,0 +1,6 @@
+{
+	"name": "Julia",
+	"image": "julia",
+	"extensions": ["julialang.language-julia"],
+	"postCreateCommand": "julia --project=. -e 'using Pkg; if (isfile(joinpath(pwd(),\"Manifest.toml\")) || isfile(joinpath(pwd(),\"JuliaManifest.toml\"))) && (isfile(joinpath(pwd(),\"Project.toml\")) || isfile(joinpath(pwd(),\"JuliaProject.toml\"))); Pkg.instantiate(); end'"
```

Comment:
Generally we're trying to keep the postCreateCommand as something developers can use as needed. These files are dropped into an existing project. Is this something that every developer would want to do with source files that already exist?

Reply:
Yes, I think this makes sense for everyone. It essentially just makes sure all the packages that are specified as a dependency in the root Julia configuration file in the repo are actually installed on the machine and available right away.

I guess in theory we could also try to put this into a docker script file, but then we would lose the ability to use the pre existing julia image files and instead would end up using building a new image for every single repo, which seems a lot less nice from a performance point of view?

We've been doing a very similar thing in the existing Julia mybinder.org integration for a long time and it seems to be what users want.

Reply:
Yeah, there's a bit of a balance between keeping it simple and getting people started that varies between ecosystems. 

Generally we've been adding Dockerfiles to help people get started on adding their own contents to the image. (For example, this [Dockerfile](https://github.com/microsoft/vscode-dev-containers/blob/master/containers/debian/.devcontainer/Dockerfile)). That's not strictly required though - it depends on how likely someone is to want to do something like this. 

One thing I did notice, however, is `git` is not present in the base Julia image. That's a pretty common dependency. There [is a script that can be used](https://github.com/microsoft/vscode-dev-containers/blob/master/script-library/common-debian.sh) that pulls in common dependencies including git. You are right that it can increase the time it takes to get the container up, however. You can default to not installing zsh and not upgrading packages to help there a bit. e.g.

```Dockerfile
# This Dockerfile adds a non-root user with sudo access. Update the “remoteUser” property in
# devcontainer.json to use it. More info: https://aka.ms/vscode-remote/containers/non-root-user.
ARG USERNAME=vscode
ARG USER_UID=1000
ARG USER_GID=$USER_UID
# Options for common setup script
ARG INSTALL_ZSH="false"
ARG UPGRADE_PACKAGES="false"
ARG COMMON_SCRIPT_SOURCE="https://raw.githubusercontent.com/microsoft/vscode-dev-containers/master/script-library/common-debian.sh"
ARG COMMON_SCRIPT_SHA="dev-mode"
# Install needed packages and setup non-root user. Use a separate RUN statement to add your own dependencies.
RUN apt-get update \
    && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install --no-install-recommends curl ca-certificates 2>&1 \
    && curl -sSL  ${COMMON_SCRIPT_SOURCE} -o /tmp/common-setup.sh \
    && ([ "${COMMON_SCRIPT_SHA}" = "dev-mode" ] || (echo "${COMMON_SCRIPT_SHA} */tmp/common-setup.sh" | sha256sum -c -)) \
    && /bin/bash /tmp/common-setup.sh "${INSTALL_ZSH}" "${USERNAME}" "${USER_UID}" "${USER_GID}" "${UPGRADE_PACKAGES}" \
    && rm /tmp/common-setup.sh \
    && apt-get autoremove -y \
    && apt-get clean -y \
    && rm -rf /var/lib/apt/lists/*
```

The package upgrade is mainly something we have for scenarios where a downstream image is actually generated so security patches are picked up before its published.

Reply:
Ah, `git` is a good point! I'll try to figure out whether there is some pre-built image for julia that includes `git`, that would give us the fastest story, right? If not, I'll look into the custom dockerfile.

Reply:
Or would it make sense that we (the Julia extension for VS Code) publish our own `julia-devcontainer` image for this scenario? That way we could include anything we think is useful, but avoid the extra time to build a new docker image every time a user instantiates a new devcontainer?

I guess this is a scenario where https://github.com/microsoft/vscode-remote-release/issues/3441 would also help: you (the codespace implementation) could actually cache the docker image that gets build by a particular devcontainer definition that is in the central repo and then reuse that docker image for any user instance that references it. mybinder.org has a pretty neat design where various docker images get cached at various points, there might be some good ideas there.

Reply:
Publishing your own image makes a ton of sense since that gives you control over when updates need to happen to the image. There's a number of other examples of this in the repository - e.g.  Puppet, Salesforce, Azure Functions. 

The thing about microsoft/vscode-remote-release#3441 that is not covered by the image alone is the various `devcontainer.json` settings. In some cases these can be more involved, so we've considered the idea of a "inheritance" for it where you can start from a definition and then override.

Reply:
Ok, the custom image works great. I just need to figure out how we can update it regularly and make those kinds of aspects a bit more stable, and then I'll update this PR here. Thanks for all the help and guidance!